ownCloud/Settings/Passcode/AppLockManager.swift

```diff
@@ -263,4 +273,30 @@ class AppLockManager: NSObject {
 			}
 		}
 	}
+
+    // MARK: - Biometrical Unlock
+
+    func authenticateWithBiometrical() {
+
+        if  shouldDisplayLockscreen, biometricalSecurityEnabled {
+
+            let context = LAContext()
+
+            // Check if the device can evaluate the policy.
+            if context.canEvaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, error: nil) {
+                context.evaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, localizedReason: "Unlock passcode".localized) { (success, _) in
+                    if success {
+                        //Fill the passcode dots
+                        DispatchQueue.main.async {
+                            self.passcodeViewController?.passcode = self.passcode
+                        }
+                        //Remove the passcode after small delay to give user feedback after use the biometrical unlock
+                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
+                            self.dismissLockscreen(animated: true)
```

Comment:
Please use `attemptUnlock()` instead, which takes care of resetting/advancing internal state on retries, too.

Reply:
👍